internal/dag/builder.go

```diff
@@ -38,53 +32,63 @@ type RouteServiceName struct {
 	Port      int32
 }
 
+// Processor uses a DAG builder to construct part
+// of a DAG.
+type Processor interface {
+	// SetBuilder provides the Processor with a Builder
+	// to use.
+	SetBuilder(builder *Builder)
+
+	// Run executes the processor.
+	Run()
+}
+
 // Builder builds a DAG.
 type Builder struct {
-
 	// Source is the source of Kubernetes objects
 	// from which to build a DAG.
 	Source KubernetesCache
 
-	// DisablePermitInsecure disables the use of the
-	// permitInsecure field in HTTPProxy.
-	DisablePermitInsecure bool
-
-	services map[RouteServiceName]*Service
+	// Processors is the ordered list of Processors to
+	// use to build the DAG.
+	Processors []Processor
 
+	services           map[RouteServiceName]*Service
 	virtualhosts       map[string]*VirtualHost
 	securevirtualhosts map[string]*SecureVirtualHost
-
-	orphaned map[types.NamespacedName]bool
-
-	FallbackCertificate *types.NamespacedName
+	listeners          []*Listener
 
 	StatusWriter
 	logrus.FieldLogger
 }
 
-// Build builds a new DAG.
+// Build builds and returns a new DAG.
 func (b *Builder) Build() *DAG {
 	b.reset()
 
-	// setup secure vhosts if there is a matching secret
-	// we do this first so that the set of active secure vhosts is stable
-	// during computeIngresses.
-	b.computeSecureVirtualhosts()
+	for _, p := range b.Processors {
+		p.SetBuilder(b)
+		p.Run()
+	}
 
-	b.computeIngresses()
+	var dag DAG
 
-	b.computeHTTPProxies()
+	// By definition, only listeners can be roots.
```

Comment:
This is not true for ext_auth clusters, which will be roots, so we need to refactor towards a model that will let the processors take care of which objects are roots (e.g. rewrite the DAG to lift hosts into listeners).

Reply:
I removed the comment. Can do the necessary refactoring in a separate PR.